[TypeTraits] Move general-purpose metaprogramming utilities from TDataFrame's utils to ROOT TypeTraits

bluehood opened the pull request:
Here it is...any kind of feedback welcome :)

bluehood commented:
A google test is on the way

amadio commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+ *************************************************************************/
+
+#ifndef ROOT_TTypeTraits
+#define ROOT_TTypeTraits
```
Isn't this supposed to drop the extra T as well?

bluehood replied:
good catch, thanks!

bluehood commented:
A google test has been added in `$ROOTSYS/core/foundation/test`. `roottest_dataframe_functiontraits` has been broken by moving the meta utils away from TDataFrame, and is removed by PR n.48 in roottest.

dpiparo commented:
@phsft-bot build please.

pcanal commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+};
+
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+struct IsV7Hist<ROOT::Experimental::THist<D, P, S...>> : public std::true_type {
```
Could we move these two back to an internal file, it is not quite 'general' enough :) [ i.e. ROOT implementation detail]

bluehood replied:
we don't want ROOT type traits in ROOT/TypeTraits? :) note that it does not introduce a dependency on libHisto, it just needs a forward decl

bluehood replied:
moved back to TDataFrame

pcanal commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
+template <typename T>
+struct TakeFirstParameter {
```
What is the difference between TakeFirst and TakeFirstParameter?  Do we need both?  Can we distinguish them more?

bluehood replied:
`TakeFirst` acts on variadic lists of types, e.g. `TakeFirst_t<A,B>` is `A`. `TakeFirstParameter` acts on the parameter list of a template object, e.g. `TakeFirstParameter_t<U<A,B>>` is `A`. They do different things, so I would say we "need" them both. I'm open to suggestions regarding the names (`TakeFirstType` and `TakeFirstParameter`?). I thought the docs were clear but maybe not?

bluehood replied:
renamed to `TakeFirstType`

pcanal replied:
I suppose your renamed the first one (formely TakeFirst), in which case, this is indeed better.  thanks

pcanal commented on core/foundation/test/TypeTraits.cxx:
```diff
@@ -0,0 +1,143 @@
+#include "ROOT/TypeTraits.hxx"
```
Can we please rename the test file either TypeTraits.test.cxx (or testTypeTrait.cxx) to distinguish it (beyhond just the directory) from an implementation file?  Thanks.

bluehood replied:
sure

vgvassilev replied:
We were using EntityToTest`Tests.cxx` pattern. Having something/test/test*.cxx looks very weird...

Axel-Naumann replied:
Who's "we"? `ls */*/test/*.cxx` seems to disagree...

Axel-Naumann commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+struct CallableTraits {
+   using arg_types = typename CallableTraits<decltype(&T::operator())>::arg_types;
+   using arg_types_nodecay = typename CallableTraits<decltype(&T::operator())>::arg_types_nodecay;
+   using ret_type = typename CallableTraits<decltype(&T::operator())>::ret_type;
```
Let me rephrase:
```
struct foo{
  int operator()(int);
  double operator()(double);
};
using res = typename CallableTraits<foo(float)>::ret_type;
```
should `res` be `double` or should this be ill-formed?

bluehood replied:
What is the standard wording? `result_of` and `invoke_result` contain a `type`, but I think it would be a confusing name here? boost uses `return_type` in their `function_traits`, I can change it to that if you prefer.

Your example should be ill-formed because `CallableTraits` does not require to specify the argument types, users just pass in a callable object. Supporting the syntax you used might be a nice addition but we have no use for it, at least for now.

Axel-Naumann replied:
Forget what I said - you test the callable, not the expression. Return type is just fine.

Axel-Naumann commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+/// Return first of possibly many template parameters.
+/// For non-template types, the result is the type itself.
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
```
That is a very ... rare ... behavior. Providing this as a general tool might see misuses / misunderstanding of the interface. I'd naively expect that `TakeFirstParameter<T>` is either empty or ill-formed, instead of mixing template levels... Do you expect generic usage of this? (Do you really know you need this behavior in your use case? I didn't look at it, it seems surprising...)

bluehood replied:
I agree, I will change the behaviour

bluehood replied:
done

Axel-Naumann commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+
+/// Compile-time integer sequence generator
+/// e.g. calling GenStaticSeq<3>::type() instantiates a StaticSeq<0,1,2>
+// TODO deprecate as soon as std::index_sequence is backported
```
Can we move this into `Experimental::`? Then we can just remove it without deprecation, and for now, no other interfaces should be using it.

bluehood replied:
I am never sure when it comes to ROOT namespaces. Do you mean `ROOT::Experimental::` or `ROOT::TypeTraits::Experimental::` or `ROOT::Experimental::TypeTraits::`?

pcanal replied:
Actually since it is not meant to be used by users (std::index_sequence is), I would put it in ::ROOT::Internal (and move it back to DataFrame unless there is another 'current' user ...

bluehood replied:
I'll move it back and push for having a backport of std::index_sequence soon :)

bluehood commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+   using Test_t = typename std::decay<T>::type;
+
+   template <typename A>
+   static constexpr bool Test(A *pt, A const *cpt = nullptr, decltype(pt->begin()) * = nullptr,
```
Thanks for looking into this. This was implemented by @dpiparo so I would ask his opinion too.

dpiparo replied:
Hi @bluehood , @Axel-Naumann I am of course fine with the new implementation!

bluehood commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+
+/// Check whether a histogram type is a classic or v7 histogram.
+template <typename T>
+struct IsV7Hist : public std::false_type {
```
@pcanal expressed the same concern. I'm outnumbered, so I'll move `IsV7Hist` back inside `TDataFrame`'s stuff.

bluehood replied:
done

bluehood commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+
+// mutable lambdas and functor classes
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...)> {
```
Can you give me an example?

This works:
```
struct A { void foo() {} };
static_assert(std::is_same<CallableTraits<decltype(&A::foo)>::ret_type, void>::value, "");
```

Axel-Naumann replied:
Does `static_assert(std::is_same<CallableTraits<decltype(A::foo)>::ret_type, void>::value, "");` work?

bluehood replied:
It does not. Can you ever use `decltype(A::foo)` anywhere?

Axel-Naumann replied:
Indeed, thanks! Forget this comment, too! :-)

bluehood commented on core/foundation/inc/ROOT/TypeTraits.hxx:
```diff
@@ -0,0 +1,203 @@
+};
+
+/// Extract types from the signature of a callable object.
+template <typename T>
```
will do

bluehood commented:
Addressed most of the comments, cleaned up history (for some reason github does not show it as it is on file).

TODO:
- make `CallableTraits` SFINAE-friendly (thanks Axel)
- remove `type` from the most generic `TakeFirstParameter`, have `TDataFrame` use something else

bluehood commented:
@Axel-Naumann sorry I can't seem to produce a reasonable use-case for the sfinae-friendliness of `CallableTraits`, could you please provide a little example snippet of how you would like it to work?

Axel-Naumann commented:
@bluehood > reasonable use-case for the sfinae-friendliness of `CallableTraits`

What about `TMVA::AddDataSource("name", T)`. That `T` is either an array of `ELEM`, or something returning `ELEM`. Would be nice if I can use `CallableTraits` to figure out the return type - if it's a function. Instead of having to hide this in another template that is then sfinae protecting the "o but your array has no call operator!" error.

As I said during our chat: I don't know whether that's a sufficiently strong reason to make `CallableTraits` sfinae friendly. But as you're publishing this to the world now you could as well make it super nice ;-)

bluehood commented:
@Axel-Naumann I am not able to write a sfinae-friendly version of `CallableTraits`. As far as I understand, your proposed usage would be the following:

```c++
template<typename T, std::size_t N>           
void foo(std::array<T,N> a) { return; }
                                                                                                                        
template<typename F, typename R = typename CallableTraits<F>::ret_type>  
void foo(F f) { return; }                       
                                                                                                                        
int main() {                                                                                                                       
   foo(std::array<int,1>{1}); // overload #2 is SFINAE'd out
   foo([]() { return std::array<int,1>{1}; }); // overload #1 does not apply                                       
   return 0;                                             
}    
```

The self-contained snippet [attached](https://github.com/root-project/root/files/1082588/testcallabletraits.txt) was my playground to try and make this work, with no success.

In principle I would like to have an empty `CallableTraits` as the most generic object, and then partially specialise for the types `T` that have `T::operator()`, but afaik it's not possible to partially specialise on properties of a type. The second idea was writing two mutually exclusive (SFINAE-wise) `template<typename T> CallableTraits`, an empty one for all types without `operator()`, and the usual one for types with `operator()`, but I don't know how to write a SFINAE expression that answers the question "does T have operator()?" with a boolean.

It's very possible that this would be trivial with concepts. Any suggestion is super welcome.

bluehood commented:
P.S. there is a [google groups conversation](https://groups.google.com/forum/#!msg/comp.lang.c++.moderated/T3x6lvmvvkQ/mfY5VTDJ--UJ) that points to an article in russian where something like this has apparently been implemented....

Axel-Naumann commented:
https://wandbox.org/permlink/vEWSUerAwZpQCdIm

bluehood commented:
well damn :)

bluehood commented:
The last two items on my to-do list have been addressed. Branch has been rebased on master to avoid failing on newly added tests.

bluehood commented:
Rebased and fixed clang-formatting. There are no items pending afaik.

bluehood commented:
Rebased on master, resolved conflicts
